domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt

```diff
@@ -35,14 +33,36 @@ class QuestionTrainingController @Inject constructor(
    */
    fun startQuestionTrainingSession(skillIdsList: List<String>): LiveData<AsyncResult<Any?>> {
     return try {
-        val questionsList = retrieveQuestionsForSkillIds(skillIdsList)
-        questionAssessmentProgressController.beginQuestionTrainingSession(questionsList)
+        questionAssessmentProgressController.beginQuestionTrainingSession(
+          retrieveQuestionsForSkillIds(skillIdsList))
         MutableLiveData(AsyncResult.success<Any?>(null))
     } catch (e: Exception) {
       MutableLiveData(AsyncResult.failed(e))
     }
   }
 
+  private fun retrieveQuestionsForSkillIds(skillIdsList: List<String>): DataProvider<List<Question>> {
+    val questionsDataProvider = topicController.retrieveQuestionsForSkillIds(skillIdsList)
+    return dataProviders.transform(TRAINING_QUESTIONS_PROVIDER, questionsDataProvider) {
+      getFilteredQuestionsForTraining(skillIdsList, it, TOTAL_QUESTIONS_PER_TOPIC/skillIdsList.size)
+    }
+  }
+
+  // Attempts to fetch equal number of questions per skill. Removes any duplicates and limits the questions to be
+  // equal to TOTAL_QUESTIONS_PER_TOPIC questions.
+  private fun getFilteredQuestionsForTraining(
+    skillIdsList: List<String>, questionsList: List<Question>, numQuestionsPerSkill: Int): List<Question>{
+    val trainingQuestions = mutableListOf<Question>()
+    for (skillId in skillIdsList) {
+      trainingQuestions.addAll(questionsList.filter {
+        it.linkedSkillIdsList.contains(skillId) &&
```

Comment:
Is it possible with this implementation to have duplicate questions?

Reply:
No because the !trainingQuestions.contains(it) line removes duplicates right?